api/python/cellxgene_census/tests/experimental/ml/test_pytorch.py

```diff
@@ -570,6 +570,37 @@ def test__shuffle(soma_experiment: Experiment) -> None:
     assert X_values == soma_joinids
 
 
+# noinspection PyTestParametrized,DuplicatedCode
+@pytest.mark.parametrize("obs_range,var_range,X_value_gen", [(16, 1, pytorch_seq_x_value_gen)])
+def test__shuffle_chunks(soma_experiment: Experiment) -> None:
+    dp = ExperimentDataPipe(
+        soma_experiment,
+        measurement_name="RNA",
+        X_name="raw",
+        obs_column_names=["label"],
+        shuffle=True,
+        soma_chunk_size=4,
+        shuffle_chunk_count=2,
+    )
+
+    all_rows = list(iter(dp))
+
+    soma_joinids = [row[1][0].item() for row in all_rows]
+    X_values = [row[0][0].item() for row in all_rows]
+
+    # same elements
+    assert set(soma_joinids) == set(range(16))
+    # not ordered! (...with a `1/16!` probability of being ordered)
+    assert soma_joinids != list(range(16))
+    # randomizes X in same order as obs
+    # note: X values were explicitly set to match obs_joinids to allow for this simple assertion
+    assert X_values == soma_joinids
+
+    # If shuffle_chunk_count is defined, each batch should contain elements from different chunks
+    batches = [soma_joinids[i : i + 4] for i in range(0, len(all_rows), 4)]
+    assert any(max(batch) - min(batch) > 3 for batch in batches)
+
```

Comment:
I would recommend against assertions like these when randomness is part of the function. Testing randomness is pretty complex topic and I think these types of assertions make the test flakey (because of randomness) and doesn't give the reader of the test much information about what is going on.

Instead of lines `599-601`, I might recommend accessing [self.soma_chunk_iter](https://github.com/chanzuckerberg/cellxgene-census/blob/0220e06462b29219df2925831d037c34244197e1/api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py#L256) of the [_ObsAndXIterator](https://github.com/chanzuckerberg/cellxgene-census/blob/0220e06462b29219df2925831d037c34244197e1/api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py#L605) object and checking:

1. `assert len(soma_chunk_iter) == obs_range // (soma_chunk_size * shuffle_chunk_count)`
2. Each element of `soma_chunk_iter` is of size `(soma_chunk_size * shuffle_chunk_count)` _except possibly the last element_
3. _Nice to have:_ If `batch_size > 1` and `batch_size` does not evenly divide `soma_chunk_size` , then check that a `batch_size` length of `soma_joinids` or `X_values` in `lines 588-589` correctly cross the boundary of `soma_chunk_iter`

(3) is probably not necessary right now but I think (1) and (2) tests the crux of the algorithm and assume randomizing the concatenated chunks just work.

I agree that doesn't test how random the scatter gather algorithm is and that is definitely worth testing but I think it involves more thought and should be a separate test altogether.

IMHO, tests of randomness are more like _stress tests_ than unit tests and they require quite a large amount of data and will _probably be slower than what is acceptable for unit tests_